src/pcap/client.go

```diff
@@ -0,0 +1,259 @@
+	logger := zap.L().With(zap.String(LogKeyHandler, "HandleRequest"))
+	// setup output/pcap-file
+	packetWriter := pcapgo.NewWriter(c.packetFile)
+	err := packetWriter.WriteFileHeader(options.SnapLen, layers.LinkTypeEthernet)
+	if err != nil {
+		return err
+	}
+
+	captureRequest := &CaptureRequest{
+		Operation: &CaptureRequest_Start{
+			Start: &StartCapture{
+				Request: endpointRequest,
+				Options: options,
+			},
+		},
+	}
+
+	c.stream, err = c.Capture(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = c.stream.Send(captureRequest)
+	if err != nil {
+		return err
+	}
+
+	copyWg := &sync.WaitGroup{}
+	copyWg.Add(1)
+	go handleStream(c.stream, packetWriter, copyWg, cancel)
+
+	go c.logProgress(ctx)
+
+	// wait for progress to finish
+	<-ctx.Done()
+
+	logger.Info("waiting for copy operation to stop")
+	copyWg.Wait()
+
+	logger.Debug("syncing file to disk")
+	err = c.packetFile.Sync()
+	if err != nil {
+		return err
+	}
+
+	logger.Debug("closing file")
+	err = c.packetFile.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Client) StopRequest() {
+	err := c.stream.SendMsg(makeStopRequest())
+	if err != nil {
+		zap.L().Panic("could not stop")
+	}
+}
+
+// handleStream reads CaptureResponse's from the api in a loop and delegates writing/logging messages & packets to writeMessage / writePacket.
+//
+// It terminates if an error or clean stop-message is received.
+func handleStream(stream API_CaptureClient, packetWriter *pcapgo.Writer, copyWg *sync.WaitGroup, cancel CancelCauseFunc) {
+	logger := zap.L().With(zap.String(LogKeyHandler, "handleStream"))
+	for {
+		res, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			logger.Info("clean stop, done")
+			cancel(nil)
+			break
+		}
+		code := status.Code(err)
+		if code != codes.OK {
+			err = fmt.Errorf("receive non-OK code: %v: %w", code, err)
+			cancel(err)
+			break
+		}
+
+		switch p := res.Payload.(type) {
+		case *CaptureResponse_Message:
+			writeMessage(p.Message)
+		case *CaptureResponse_Packet:
+			writePacket(p.Packet, packetWriter)
+		}
+	}
+	copyWg.Done()
+}
+
+// writeMessage accepts a Message and writes a log-line using a log-level corresponding to the severity of the message.
+func writeMessage(message *Message) {
+	var logLevel zapcore.Level
+
+	switch message.Type {
+	case MessageType_UNKNOWN:
+		logLevel = zapcore.WarnLevel
+	case MessageType_INSTANCE_UNAVAILABLE:
+		logLevel = zapcore.WarnLevel
+	case MessageType_START_CAPTURE_FAILED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_INVALID_REQUEST:
+		logLevel = zapcore.ErrorLevel
+	case MessageType_CONGESTED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_LIMIT_REACHED:
+		logLevel = zapcore.WarnLevel
+	case MessageType_CAPTURE_STOPPED:
+		logLevel = zapcore.InfoLevel
+	case MessageType_CONNECTION_ERROR:
+		logLevel = zapcore.ErrorLevel
+	}
+	// TODO: Find a way to assign the proper log level.
+	_ = logLevel
+	zap.S().Infof("%s(%s): %s", message.Origin, message.Type, message.Message)
```

Comment:
We agreed do not use the sugared logger.

Reply:
This is for use in the CLI, different rules apply.
But realistically, the `Client` should not log anything itself. I proposed some 'callback handler' that gets packets or messages and can decide what to do with those. The callback handler would be defined in the app that uses the client. Then that app can define how to log / format outputs.